src/ClientModel/SamplingDataPostProcessorTest.cpp

```diff
@@ -72,6 +72,75 @@ MATCHER_P(SampledFunctionEq, that, "") {
+    return std::make_tuple(lhs.timestamp_ns(), lhs.callstack_id(), lhs.thread_id()) <
+           std::make_tuple(rhs.timestamp_ns(), rhs.callstack_id(), rhs.thread_id());
+  };
+  auto callstack_events_equal = [](const CallstackEvent& lhs, const CallstackEvent& rhs) {
+    return std::make_tuple(lhs.timestamp_ns(), lhs.callstack_id(), lhs.thread_id()) ==
+           std::make_tuple(rhs.timestamp_ns(), rhs.callstack_id(), rhs.thread_id());
+  };
+
+  std::vector<CallstackEvent> lhs_events = lhs.second;
+  std::sort(lhs_events.begin(), lhs_events.end(), callstack_events_less);
+  std::vector<CallstackEvent> rhs_events = rhs.second;
+  std::sort(rhs_events.begin(), rhs_events.end(), callstack_events_less);
+  return lhs.first == rhs.first &&
+         std::equal(lhs_events.begin(), lhs_events.end(), rhs_events.begin(), rhs_events.end(),
+                    callstack_events_equal);
+};
+
+MATCHER_P(CallstackIdToCallstackEventsEq, that, "") {
+  const std::pair<uint64_t, std::vector<CallstackEvent>>& lhs = arg;
+  const std::pair<uint64_t, std::vector<CallstackEvent>>& rhs = that;
+  return CallstackIdToCallstackEventPairsAreEqual(lhs, rhs);
+}
+
+MATCHER_P(ThreadSampleDataEq, that, "") {
+  const ThreadSampleData& lhs = arg;
+  const ThreadSampleData& rhs = that;
+
+  // Compare sampled_callstack_id_to_events.
+  auto sampled_callstack_id_to_event_pair_less =
+      [](const std::pair<uint64_t, std::vector<CallstackEvent>>& l,
+         const std::pair<uint64_t, std::vector<CallstackEvent>>& r) { return l.first < r.first; };
+
+  std::vector<std::pair<uint64_t, std::vector<CallstackEvent>>>
+      lhs_sampled_callstack_id_to_event_pairs;
+  for (const auto& [callstack_id, events] : lhs.sampled_callstack_id_to_events) {
+    lhs_sampled_callstack_id_to_event_pairs.emplace_back(callstack_id, events);
+  }
+  std::sort(lhs_sampled_callstack_id_to_event_pairs.begin(),
+            lhs_sampled_callstack_id_to_event_pairs.end(), sampled_callstack_id_to_event_pair_less);
+  std::vector<std::pair<uint64_t, std::vector<CallstackEvent>>>
+      rhs_sampled_callstack_id_to_event_pairs;
+  for (const auto& [callstack_id, events] : rhs.sampled_callstack_id_to_events) {
+    rhs_sampled_callstack_id_to_event_pairs.emplace_back(callstack_id, events);
+  }
+  std::sort(rhs_sampled_callstack_id_to_event_pairs.begin(),
+            rhs_sampled_callstack_id_to_event_pairs.end(), sampled_callstack_id_to_event_pair_less);
+  if (!std::equal(lhs_sampled_callstack_id_to_event_pairs.begin(),
+                  lhs_sampled_callstack_id_to_event_pairs.end(),
+                  rhs_sampled_callstack_id_to_event_pairs.begin(),
+                  rhs_sampled_callstack_id_to_event_pairs.end(),
+                  CallstackIdToCallstackEventPairsAreEqual)) {
+    return false;
+  }
+
+  // Compare the rest.
+  return lhs.thread_id == rhs.thread_id && lhs.samples_count == rhs.samples_count &&
```

Comment:
It doesn't matter as this is only for tests. But usually, I would prefer to always first compare the constant-time members before more expensive comparisons. That way you can early exit and have a fast "not equal" response.

But as said, completely irrelevant in the case of this test.

Reply:
It turns out that without your comment I realized that this method is not used at all after the refactorings from the previous days, that made `PostProcessedSamplingData` return pointers to `ThreadSampleData`... I'll remove it.

Reply:
Edit: with

Reply:
Removed this ugly block of code.

Reply:
Thanks, lgtm!